File: ExtremeStudio/DockPanel Forms/MainForm Docks/ObjectExplorerDock/ObjectExplorerDock.vb
Imports System.IO
Imports System.Text.RegularExpressions
Imports ExtremeParser
Imports ScintillaNET

Public Class ObjectExplorerDock
    Dim _nodeState As ExtremeCore.treeNodeStateSaving = New ExtremeCore.treeNodeStateSaving

#Region "Funcs"
    Private Function GetTreeViewPart(parser As ExtremeParser.CodeParts, searchTerm As String, tag as String)
        If parser Is Nothing Then Return Nothing

        Dim mainNode As New TreeNode With {
            .Text = Path.GetFileNameWithoutExtension(parser.FilePath)
        }

        Dim defines As TreeNode
        Dim macros As TreeNode
        Dim functions As TreeNode
        Dim publics As TreeNode
        Dim stocks As TreeNode
        Dim natives As TreeNode

        If parser.Defines.FindAll(Function(x) x.DefineName.Contains(searchTerm)).Count > 0 Then defines = mainNode.Nodes.Add("Defines")
        If parser.Macros.FindAll(Function(x) x.DefineName.Contains(searchTerm)).Count > 0 Then macros = mainNode.Nodes.Add("Macros")
        If parser.Functions.FindAll(Function(x) x.FuncName.Contains(searchTerm)).Count > 0 Then functions = mainNode.Nodes.Add("Functions")
        If parser.Publics.FindAll(Function(x) x.FuncName.Contains(searchTerm)).Count > 0 Then publics = mainNode.Nodes.Add("Publics")
        If parser.Stocks.FindAll(Function(x) x.FuncName.Contains(searchTerm)).Count > 0 Then stocks = mainNode.Nodes.Add("Stocks")
        If parser.Natives.FindAll(Function(x) x.FuncName.Contains(searchTerm)).Count > 0 Then natives = mainNode.Nodes.Add("Natives")

        'Create the custom Roots.
        Dim listCustom As New List(Of TreeNode)
        For Each itm In MainForm.currentProject.objectExplorerItems
            Dim bla = mainNode.Nodes.Add(itm.Name) : bla.Tag = itm.Identifier 'Here I set its tag to the identifer temporarly, It will be changed to `Root` again in Functions loop.
            listCustom.Add(bla)
        Next

        'Start
        For Each key In parser.Defines.FindAll(Function(x) x.DefineName.Contains(searchTerm))
            Dim nde = defines.Nodes.Add(key.DefineName)
            nde.ToolTipText = "Define Value: " + vbCrLf + key.DefineValue
            nde.Tag = "define|" + tag
        Next

        For Each key In parser.Macros.FindAll(Function(x) x.DefineName.Contains(searchTerm))
            Dim nde = macros.Nodes.Add(key.DefineName)
            nde.ToolTipText = "Define Value: " + vbCrLf + key.DefineValue
            nde.Tag = "define|" + tag
        Next

        For Each funcs In parser.Functions.FindAll(Function(x) x.FuncName.Contains(searchTerm))
            Dim done As Boolean = False

            'Check if it crosponds to a custom one first.
            For Each itm In listCustom
                If funcs.FuncName.StartsWith(itm.Tag) Then
                    Dim node = itm.Nodes.Add(funcs.FuncName)
                    node.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, funcs).ToolTipText
                    done = True 'To skip the `Else if it wasn't used.`
                    node.Tag = "function|" + tag
                    Exit For
                End If
            Next

            If done = True Then Continue For

            'Else if it wasn't used.
            Dim nde = functions.Nodes.Add(funcs.FuncName)
            nde.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, funcs).ToolTipText
            nde.Tag = "function|" + tag
        Next

        For Each publicFunc In parser.Publics.FindAll(Function(x) x.FuncName.Contains(searchTerm))
            Dim done As Boolean = False

            'Check if it crosponds to a custom one first.
            For Each itm In listCustom
                If publicFunc.FuncName.StartsWith(itm.Tag) Then
                    Dim node = itm.Nodes.Add(publicFunc.FuncName)
                    node.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, publicFunc).ToolTipText
                    done = True 'To skip the `Else if it wasn't used.`
                    node.Tag = "public|" + tag
                    Exit For
                End If
            Next

            If done = True Then Continue For

            'Else if it wasn't used.
            Dim nde = publics.Nodes.Add(publicFunc.FuncName)
            nde.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, publicFunc).ToolTipText
            nde.Tag = "public|" + tag
        Next

        For Each stock In parser.Stocks.FindAll(Function(x) x.FuncName.Contains(searchTerm))
            Dim done As Boolean = False

            'Check if it crosponds to a custom one first.
            For Each itm In listCustom
                If stock.FuncName.StartsWith(itm.Tag) Then
                    Dim node = itm.Nodes.Add(stock.FuncName)
                    node.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, stock).ToolTipText
                    done = True 'To skip the `Else if it wasn't used.`
                    node.Tag = "stock|" + tag
                    Exit For
                End If
            Next

            If done = True Then Continue For

            'Else if it wasn't used.
            Dim nde = stocks.Nodes.Add(stock.FuncName)
            nde.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, stock).ToolTipText
            nde.Tag = "stock|" + tag
        Next

        For Each native In parser.Natives.FindAll(Function(x) x.FuncName.Contains(searchTerm))
            Dim done As Boolean = False

            'Check if it crosponds to a custom one first.
            For Each itm In listCustom
                If native.FuncName.StartsWith(itm.Tag) Then
                    Dim node = itm.Nodes.Add(native.FuncName)
                    node.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, native).ToolTipText
                    done = True 'To skip the `Else if it wasn't used.`
                    node.Tag = "native|" + tag
                    Exit For
                End If
            Next

            If done = True Then Continue For

            'Else if it wasn't used.
            Dim nde = natives.Nodes.Add(native.FuncName)
            nde.ToolTipText = New AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, native).ToolTipText
            nde.Tag = "native|" + tag
        Next
        Return mainNode
    End Function

    Public Sub RefreshTreeView(parts As CodeParts)
        'Save.
        _nodeState.SaveTreeState(treeView.Nodes)

        'Clear All.
        treeView.Nodes.Clear()

        'Start by adding the current file.
        Dim node As TreeNode = getTreeViewPart(parts, SearchTextBox.Text, "current")
        node.Text = "Current File"
        If node.Nodes.Count > 0 Then treeView.Nodes.Add(node)

        'Loop through all includes and put them.
        Dim parentNode As New TreeNode With {
            .Text = "Includes"
        }
        Dim allIncs = parts.FlattenIncludes
        For i As Integer = 1 To allIncs.Count - 1 'Loop through all skipping ID 0.
            Dim nds = getTreeViewPart(allIncs(i), SearchTextBox.Text, allIncs(i).FilePath)
            If nds.Nodes.Count > 0 Then parentNode.Nodes.Add(nds)
        Next

        'Add to list.
        If parentNode.Nodes.Count > 0 Then treeView.Nodes.Add(parentNode)

        _nodeState.RestoreTreeState(treeView) 'Restore
    End Sub
#End Region

    Private Sub ObjectExplorerDock_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If MainForm.CurrentEditor IsNot Nothing Then
            refreshTreeView(MainForm.CurrentEditor.codeParts)
        End If
    End Sub

    Private Sub EditItemsToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles EditItemsToolStripMenuItem.Click
        ObjectExplorerDockItems.ShowDialog()
    End Sub

    Private Sub ToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles ToolStripMenuItem1.Click
        refreshTreeView(MainForm.CurrentEditor.codeParts)
    End Sub

    Private Sub SearchTextBox_Changed(sender As Object, e As EventArgs) Handles SearchTextBox.TextChanged
        If SearchTextBox.Text = "" Then
            Label1.Visible = True
        Else
            Label1.Visible = False
        End If
    End Sub

    Private Sub Label1_Click(sender As Object, e As EventArgs) Handles Label1.Click
        SearchTextBox.Focus()
    End Sub

    Private Sub SearchTextBox_KeyPress(sender As Object, e As KeyPressEventArgs) Handles SearchTextBox.KeyPress
        If e.KeyChar = ChrW(Keys.Return) Then
            refreshTreeView(MainForm.CurrentEditor.codeParts)
            e.Handled = True
        End If
    End Sub

    Private Sub FindAndGoto(rgx As string)
        If Mainform.CurrentScintilla Is Nothing Then Exit Sub
        Dim match = Regex.Match(Mainform.CurrentScintilla.GetTextRange(0, Mainform.CurrentScintilla.TextLength), rgx, RegexOptions.Multiline)
        If match.Index <> 0 And match.Length <> 0 Then
            MainForm.CurrentScintilla.SetSelection(match.Index, match.Index + match.Length)
            Mainform.CurrentScintilla.ScrollCaret()
        End If
        'I could have just used ScintillaNEts regex here but it for some reason doesn't provide multiline regex which is needed for accurancy.
    End Sub

    Private Sub treeView_NodeMouseDoubleClick(sender As Object, e As TreeNodeMouseClickEventArgs) Handles treeView.NodeMouseDoubleClick
        If e.Node.Tag = Nothing Then Exit Sub

        Dim filepath As String = e.Node.Tag.Split("|")(1)
        If filepath <> "current" Then
            'First, Loop throuhg all opened tabs and see if its already opened there.
            Dim isFound as Boolean = False
            For Each doc In Mainform.MainDock.Documents
                If doc.DockHandler.Form.Controls("Editor").Tag = filepath Then
                    doc.DockHandler.Activate()
                    isFound = True
                End If
            Next

            If isFound = False Then
                'If not found, Open that file.
                Mainform.OpenFile(filepath)
            End If
        End If

        If e.Node.Tag.StartsWith("define") Then
            FindAndGoto("^[ \t]*[#]define[ \t]+" + Regex.Escape(e.Node.Text) + "[ \t]+(?:\\\s+)?(?>(?<value>[^\\\n\r]+)[ \t]*(?:\\\s+)?)*")
        ElseIf e.Node.Tag.StartsWith("function") Then
            FindAndGoto("^[ \t]*(?:\sstatic\s+stock\s+|\sstock\s+static\s+|\sstatic\s+)?" + Regex.Escape(e.Node.Text) + "\((.*)\)(?!;)\s*{")
        ElseIf e.Node.Tag.StartsWith("public") Then
            FindAndGoto("public[ \t]+" + Regex.Escape(e.Node.Text) + "[ \t]*\((.*)\)\s*{")
        ElseIf e.Node.Tag.StartsWith("stock") Then
            FindAndGoto("stock[ \t]+" + Regex.Escape(e.Node.Text) + "[ \t]*\((.*)\)\s*{")
        ElseIf e.Node.Tag.StartsWith("native") Then
            FIndAndGoto("native[ \t]+" + Regex.Escape(e.Node.Text) + "[ \t]*?\((.*)\);")
        End If
    End Sub
End Class

File: ExtremeStudio/DockPanelForms/MainFormDocks/ObjectExplorerDock/ObjectExplorerDock.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows.Forms;
using ExtremeCore.Classes;
using ExtremeParser;
using ExtremeStudio.DockPanelForms.MainFormDocks.EditorDock.AutoComplete;
using WeifenLuo.WinFormsUI.Docking;

namespace ExtremeStudio.DockPanelForms.MainFormDocks.ObjectExplorerDock
{
    public partial class ObjectExplorerDock : DockContent
    {
        public ObjectExplorerDock()
        {
            InitializeComponent();
        }

        TreeNodeStateSaver _nodeState = new TreeNodeStateSaver();

        #region Funcs

        [Localizable(false)]
        private dynamic GetTreeViewPart(CodeParts parser, string searchTerm, string tag)
        {
            if (ReferenceEquals(parser, null))
            {
                return null;
            }

            TreeNode mainNode = new TreeNode
            {
                Text = Path.GetFileNameWithoutExtension(Convert.ToString(parser.FilePath))
            };

            TreeNode defines = null;
            TreeNode macros = null;
            TreeNode functions = null;
            TreeNode publics = null;
            TreeNode stocks = null;
            TreeNode natives = null;

            if (parser.Defines.FindAll(x => x.DefineName.Contains(searchTerm)).Any())
            {
                defines = mainNode.Nodes.Add("Defines");
            }

            if (parser.Macros.FindAll(x => x.DefineName.Contains(searchTerm)).Any())
            {
                macros = mainNode.Nodes.Add("Macros");
            }

            if (parser.Functions.FindAll(x => x.FuncName.Contains(searchTerm)).Any())
            {
                functions = mainNode.Nodes.Add("Functions");
            }

            if (parser.Publics.FindAll(x => x.FuncName.Contains(searchTerm)).Any())
            {
                publics = mainNode.Nodes.Add("Publics");
            }

            if (parser.Stocks.FindAll(x => x.FuncName.Contains(searchTerm)).Any())
            {
                stocks = mainNode.Nodes.Add("Stocks");
            }

            if (parser.Natives.FindAll(x => x.FuncName.Contains(searchTerm)).Any())
            {
                natives = mainNode.Nodes.Add("Natives");
            }

            //Create the custom Roots.
            List<TreeNode> listCustom = new List<TreeNode>();
            foreach (var itm in Program.MainForm.CurrentProject.ObjectExplorerItems)
            {
                var bla = mainNode.Nodes.Add(itm.Name);
                bla.Tag = itm
                    .Identifier; //Here I set its tag to the identifer temporarly, It will be changed to `Root` again in Functions loop.
                listCustom.Add(bla);
            }

            //Start
            foreach (var key in parser.Defines.FindAll(x => x.DefineName.Contains(searchTerm)))
            {
                var nde = defines.Nodes.Add(key.DefineName);
                nde.ToolTipText = "Define Value: " + "\r\n" + key.DefineValue;
                nde.Tag = "define|" + tag;
            }

            foreach (var key in parser.Macros.FindAll(x => x.DefineName.Contains(searchTerm)))
            {
                var nde = macros.Nodes.Add(key.DefineName);
                nde.ToolTipText = "Define Value: " + "\r\n" + key.DefineValue;
                nde.Tag = "define|" + tag;
            }

            foreach (var funcs in parser.Functions.FindAll(x => x.FuncName.Contains(searchTerm)))
            {
                bool done = false;

                //Check if it crosponds to a custom one first.
                foreach (var itm in listCustom)
                {
                    if (funcs.FuncName.StartsWith((string)itm.Tag))
                    {
                        var node = itm.Nodes.Add(funcs.FuncName);
                        node.ToolTipText =
                            new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, funcs)
                                .ToolTipText;
                        done = true; //To skip the `Else if it wasn't used.`
                        node.Tag = "function|" + tag;
                        break;
                    }
                }

                if (done)
                {
                    continue;
                }

                //Else if it wasn't used.
                var nde = functions.Nodes.Add(funcs.FuncName);
                nde.ToolTipText = new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, funcs)
                    .ToolTipText;
                nde.Tag = "function|" + tag;
            }

            foreach (var publicFunc in parser.Publics.FindAll(x => x.FuncName.Contains(searchTerm)))
            {
                bool done = false;

                //Check if it crosponds to a custom one first.
                foreach (var itm in listCustom)
                {
                    if (publicFunc.FuncName.StartsWith((string)itm.Tag))
                    {
                        var node = itm.Nodes.Add(publicFunc.FuncName);
                        node.ToolTipText =
                            new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, publicFunc)
                                .ToolTipText;
                        done = true; //To skip the `Else if it wasn't used.`
                        node.Tag = "public|" + tag;
                        break;
                    }
                }

                if (done)
                {
                    continue;
                }

                //Else if it wasn't used.
                var nde = publics.Nodes.Add(publicFunc.FuncName);
                nde.ToolTipText = new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, publicFunc)
                    .ToolTipText;
                nde.Tag = "public|" + tag;
            }

            foreach (var stock in parser.Stocks.FindAll(x => x.FuncName.Contains(searchTerm)))
            {
                bool done = false;

                //Check if it crosponds to a custom one first.
                foreach (var itm in listCustom)
                {
                    if (stock.FuncName.StartsWith((string)itm.Tag))
                    {
                        var node = itm.Nodes.Add(stock.FuncName);
                        node.ToolTipText =
                            new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, stock)
                                .ToolTipText;
                        done = true; //To skip the `Else if it wasn't used.`
                        node.Tag = "stock|" + tag;
                        break;
                    }
                }

                if (done)
                {
                    continue;
                }

                //Else if it wasn't used.
                var nde = stocks.Nodes.Add(stock.FuncName);
                nde.ToolTipText = new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, stock)
                    .ToolTipText;
                nde.Tag = "stock|" + tag;
            }

            foreach (var native in parser.Natives.FindAll(x => x.FuncName.Contains(searchTerm)))
            {
                bool done = false;

                //Check if it crosponds to a custom one first.
                foreach (var itm in listCustom)
                {
                    if (native.FuncName.StartsWith((string)itm.Tag))
                    {
                        var node = itm.Nodes.Add(native.FuncName);
                        node.ToolTipText =
                            new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, native)
                                .ToolTipText;
                        done = true; //To skip the `Else if it wasn't used.`
                        node.Tag = "native|" + tag;
                        break;
                    }
                }

                if (done)
                {
                    continue;
                }

                //Else if it wasn't used.
                var nde = natives.Nodes.Add(native.FuncName);
                nde.ToolTipText = new AutoCompleteItemEx(AutoCompleteItemEx.AutoCompeleteTypes.TypeFunction, native)
                    .ToolTipText;
                nde.Tag = "native|" + tag;
            }

            return mainNode;
        }

        [Localizable(false)]
        public void RefreshTreeView(CodeParts parts)
        {
            if (parts == null)
                return;

            //Save.
            _nodeState.SaveTreeState(treeView.Nodes);

            //Clear All.
            treeView.Nodes.Clear();

            //Start by adding the current file.
            TreeNode node = GetTreeViewPart(parts, Convert.ToString(SearchTextBox.Text), "current");
            node.Text = "Current File";
            if (node.Nodes.Count > 0)
            {
                treeView.Nodes.Add(node);
            }

            //Loop through all includes and put them.
            TreeNode parentNode = new TreeNode
            {
                Text = "Includes"
            };
            var allIncs = parts.FlattenIncludes();
            var codePartses = allIncs as CodeParts[] ?? allIncs.ToArray();
            for (int i = 1; i <= codePartses.Count() - 1; i++) //Loop through all skipping ID 0.
            {
                var nds = GetTreeViewPart(codePartses[i], Convert.ToString(SearchTextBox.Text),
                    Convert.ToString(codePartses[i].FilePath));
                if (nds.Nodes.Count > 0)
                {
                    parentNode.Nodes.Add(nds);
                }
            }

            //Add to list.
            if (parentNode.Nodes.Count > 0)
            {
                treeView.Nodes.Add(parentNode);
            }

            _nodeState.RestoreTreeState(treeView); //Restore
        }

        #endregion

        private void ObjectExplorerDock_Load(object sender, EventArgs e)
        {
            if (Program.MainForm.CurrentEditor != null)
            {
                RefreshTreeView(Program.MainForm.CurrentEditor.CodeParts);
            }
        }

        private void EditItemsToolStripMenuItem_Click(object sender, EventArgs e)
        {
            new ObjectExplorerDockItems().ShowDialog();
        }

        private void ToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            RefreshTreeView(Program.MainForm?.CurrentEditor?.CodeParts);
        }

        private void SearchTextBox_Changed(object sender, EventArgs e)
        {
            if (SearchTextBox.Text == "")
            {
                Label1.Visible = true;
            }
            else
            {
                Label1.Visible = false;
            }
        }

        private void Label1_Click(object sender, EventArgs e)
        {
            SearchTextBox.Focus();
        }

        private void SearchTextBox_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (e.KeyChar == (char)Keys.Return)
            {
                RefreshTreeView(Program.MainForm.CurrentEditor.CodeParts);
                e.Handled = true;
            }
        }

        private void FindAndGoto(string rgx)
        {
            if (ReferenceEquals(Program.MainForm.CurrentScintilla, null))
            {
                return;
            }

            var match = Regex.Match(Program.MainForm.CurrentScintilla.GetTextRange(0, Program.MainForm.CurrentScintilla.TextLength),
                rgx, RegexOptions.Multiline);
            if (match.Index != 0 && match.Length != 0)
            {
                Program.MainForm.CurrentScintilla.SetSelection(match.Index, match.Index + match.Length);
                Program.MainForm.CurrentScintilla.ScrollCaret();
            }

            //I could have just used ScintillaNEts regex here but it for some reason doesn't provide multiline regex which is needed for accurancy.
        }

        [Localizable(false)]
        private void treeView_NodeMouseDoubleClick(object sender, TreeNodeMouseClickEventArgs e)
        {
            if (e.Node.Tag == null)
            {
                return;
            }

            string filepath = Convert.ToString(((string)e.Node.Tag).Split('|')[1]);
            if (filepath != "current")
            {
                //First, Loop throuhg all opened tabs and see if its already opened there.
                bool isFound = false;
                foreach (var doc in Program.MainForm.MainDock.Documents)
                {
                    if ((string)doc.DockHandler.Form.Controls["Editor"].Tag == filepath)
                    {
                        doc.DockHandler.Activate();
                        isFound = true;
                    }
                }

                if (isFound == false)
                {
                    //If not found, Open that file.
                    Program.MainForm.OpenFile(filepath);
                }
            }

            if (((string)e.Node.Tag).StartsWith("define"))
            {
                FindAndGoto("^[ \\t]*[#]define[ \\t]+" + Regex.Escape(e.Node.Text) +
                            "[ \\t]+(?:\\\\\\s+)?(?>(?<value>[^\\\\\\n\\r]+)[ \\t]*(?:\\\\\\s+)?)*");
            }
            else if (((string)e.Node.Tag).StartsWith("function"))
            {
                FindAndGoto("^[ \\t]*(?:\\sstatic\\s+stock\\s+|\\sstock\\s+static\\s+|\\sstatic\\s+)?" +
                            Regex.Escape(e.Node.Text) + "\\((.*)\\)(?!;)\\s*{");
            }
            else if (((string)e.Node.Tag).StartsWith("public"))
            {
                FindAndGoto("public[ \\t]+" + Regex.Escape(e.Node.Text) + "[ \\t]*\\((.*)\\)\\s*{");
            }
            else if (((string)e.Node.Tag).StartsWith("stock"))
            {
                FindAndGoto("stock[ \\t]+" + Regex.Escape(e.Node.Text) + "[ \\t]*\\((.*)\\)\\s*{");
            }
            else if (((string)e.Node.Tag).StartsWith("native"))
            {
                FindAndGoto("native[ \\t]+" + Regex.Escape(e.Node.Text) + "[ \\t]*?\\((.*)\\);");
            }
        }
    }
}